### PR TITLE
Backup: `--incremental-from-pos` supports backup name

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

Running an incremental backup, up till now the flag `--incremental-from-pos` supported:

- A position
- The magic word `"auto"`

It now also supports a backup name. If such named backup exists, then the incremental backup will start with the position of the named backup.

A bit of refactoring made, especially around `endtoend` tests, to accommodate.

Docs: ~~https://github.com/vitessio/website/pull/1662~~ https://github.com/vitessio/website/pull/1667

## Related Issue(s)

Fixes https://github.com/vitessio/vitess/issues/14905

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

**Comment by shlomi-noach** [comment]
Documented in https://github.com/vitessio/website/pull/1662

File: go/vt/mysqlctl/backupengine.go [hunk 1]
```diff
@@ -446,14 +449,40 @@ func FindLatestSuccessfulBackupPosition(ctx context.Context, params BackupParams
...
+		if bh.Name() != backupName {
+			continue
+		}
+		manifest, err := GetBackupManifest(ctx, bh)
+		if err != nil {
+			return pos, vterrors.Wrapf(err, "GetBackupManifest failed for backup: %v", backupName)
+		}
+		return manifest.Position, nil
+	}
+	return pos, vterrors.Errorf(vtrpc.Code_NOT_FOUND, "GetBackupManifest could not find backup: %v", backupName)
```
Comment (mattlord) [inline go/vt/mysqlctl/backupengine.go:483]:
```suggestion
	return pos, vterrors.Errorf(vtrpc.Code_NOT_FOUND, "could not find backup %q for %s/%s", backupName, params.Keyspace, params.Shard)
```

File: go/vt/mysqlctl/builtinbackupengine.go [hunk 2]
```diff
@@ -257,25 +261,24 @@ func (be *BuiltinBackupEngine) executeIncrementalBackup(ctx context.Context, par
...
-		}
-		purgedGTIDSet, ok := gtidPurged.GTIDSet.(replication.Mysql56GTIDSet)
-		if !ok {
-			return gtidPurged, nil, vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "cannot get MySQL GTID purged value: %v", gtidPurged)
+			return false, err
 		}
-		return gtidPurged, purgedGTIDSet, nil
+		fromBackupName = backupName
+		params.IncrementalFromPos = replication.EncodePosition(pos)
+		params.Logger.Infof("auto evaluated incremental_from_pos: %s", params.IncrementalFromPos)
```
Comment (mattlord) [inline go/vt/mysqlctl/builtinbackupengine.go:273]:
```suggestion
		params.Logger.Infof("evaluated incremental_from_pos using backup name %q: %s", backupName, params.IncrementalFromPos)
```

File: go/vt/mysqlctl/builtinbackupengine.go [hunk 3]
```diff
@@ -291,6 +294,19 @@ func (be *BuiltinBackupEngine) executeIncrementalBackup(ctx context.Context, par
...
+
+	// @@gtid_purged
+	getPurgedGTIDSet := func() (replication.Position, replication.Mysql56GTIDSet, error) {
+		gtidPurged, err := params.Mysqld.GetGTIDPurged(ctx)
+		if err != nil {
+			return gtidPurged, nil, vterrors.Wrap(err, "can't get @@gtid_purged")
+		}
+		purgedGTIDSet, ok := gtidPurged.GTIDSet.(replication.Mysql56GTIDSet)
+		if !ok {
+			return gtidPurged, nil, vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "cannot get MySQL GTID purged value: %v", gtidPurged)
```
Comment (mattlord) [inline go/vt/mysqlctl/builtinbackupengine.go:306]:
```suggestion
			return gtidPurged, nil, vterrors.Errorf(vtrpc.Code_FAILED_PRECONDITION, "failed to parse a valid MySQL GTID set from value: %v", gtidPurged)
```

**Review by mattlord — COMMENTED** [review]
LGTM. Just a few minor suggestions.

At first I wasn't crazy about re-using the existing flag, but after reading through it and thinking about it I think it's fine. The backup name is a pointer to a pos, whatever pos that file had. We'll just need to document it of course here: https://vitess.io/docs/19.0/user-guides/operating-vitess/backup-and-restore/creating-a-backup/#create-an-incremental-backup-with-vtctl

The only reason that I didn't approve here, is that we should update the `--incremental-from-pos`/`--incremental_from_pos` flag related text:
  - https://github.com/vitessio/vitess/blob/main/go/cmd/vtctldclient/command/backups.go
  - https://github.com/vitessio/vitess/blob/main/go/vt/vtctl/backup.go
  - https://github.com/vitessio/vitess/blob/main/go/cmd/vtbackup/cli/vtbackup.go

Unless I'm missing something here?

**Comment by shlomi-noach** [comment]
> At first I wasn't crazy about re-using the existing flag, but after reading through it and thinking about it I think it's fine.

I think ideally the flag should be renamed `--incremental-from`, and then it makes a ton of sense. I can work on transitioning the flag through two versions, in a followup PR.

**Comment by shlomi-noach** [comment]
Updated all CLI docs.

**Review by mattlord — APPROVED** [review]
Thanks, @shlomi-noach ! ❤️

**Comment by shlomi-noach** [comment]
New docs PR in https://github.com/vitessio/website/pull/1667, cleaner.

File: go/cmd/vtbackup/cli/vtbackup.go [hunk 4]
```diff
@@ -204,7 +204,7 @@ func init() {
 	Main.Flags().StringVar(&initKeyspace, "init_keyspace", initKeyspace, "(init parameter) keyspace to use for this tablet")
 	Main.Flags().StringVar(&initShard, "init_shard", initShard, "(init parameter) shard to use for this tablet")
 	Main.Flags().IntVar(&concurrency, "concurrency", concurrency, "(init restore parameter) how many concurrent files to restore at once")
-	Main.Flags().StringVar(&incrementalFromPos, "incremental_from_pos", incrementalFromPos, "Position of previous backup. Default: empty. If given, then this backup becomes an incremental backup from given position. If value is 'auto', backup taken from last successful backup position")
+	Main.Flags().StringVar(&incrementalFromPos, "incremental_from_pos", incrementalFromPos, "Position, or name of backup from which to create an incremental backup. Default: empty. If given, then this backup becomes an incremental backup from given position or given backup. If value is 'auto', backup taken from last successful backup position.")
```
Comment (deepthi) [inline go/cmd/vtbackup/cli/vtbackup.go:207]:
```suggestion
	Main.Flags().StringVar(&incrementalFromPos, "incremental_from_pos", incrementalFromPos, "Position, or name of backup from which to create an incremental backup. Default: empty. If given, then this backup becomes an incremental backup from given position or given backup. If value is 'auto', this backup will be taken from the last successful backup position.")
```

File: go/cmd/vtctldclient/command/backups.go [hunk 5]
```diff
@@ -281,14 +281,14 @@ func commandRestoreFromBackup(cmd *cobra.Command, args []string) error {
 func init() {
 	Backup.Flags().BoolVar(&backupOptions.AllowPrimary, "allow-primary", false, "Allow the primary of a shard to be used for the backup. WARNING: If using the builtin backup engine, this will shutdown mysqld on the primary and stop writes for the duration of the backup.")
 	Backup.Flags().Int32Var(&backupOptions.Concurrency, "concurrency", 4, "Specifies the number of compression/checksum jobs to run simultaneously.")
-	Backup.Flags().StringVar(&backupOptions.IncrementalFromPos, "incremental-from-pos", "", "Position of previous backup. Default: empty. If given, then this backup becomes an incremental backup from given position. If value is 'auto', backup taken from last successful backup position")
+	Backup.Flags().StringVar(&backupOptions.IncrementalFromPos, "incremental-from-pos", "", "Position, or name of backup from which to create an incremental backup. Default: empty. If given, then this backup becomes an incremental backup from given position or given backup. If value is 'auto', backup taken from last successful backup position")
```
Comment (deepthi) [inline go/cmd/vtctldclient/command/backups.go:284]:
```suggestion
	Backup.Flags().StringVar(&backupOptions.IncrementalFromPos, "incremental-from-pos", "", "Position, or name of backup from which to create an incremental backup. Default: empty. If given, then this backup becomes an incremental backup from given position or given backup. If value is 'auto', this backup will be taken from the last successful backup position.")
```

File: go/cmd/vtctldclient/command/backups.go [hunk 6]
```diff
@@ -281,14 +281,14 @@ func commandRestoreFromBackup(cmd *cobra.Command, args []string) error {
...
 	Root.AddCommand(Backup)
 
 	BackupShard.Flags().BoolVar(&backupShardOptions.AllowPrimary, "allow-primary", false, "Allow the primary of a shard to be used for the backup. WARNING: If using the builtin backup engine, this will shutdown mysqld on the primary and stop writes for the duration of the backup.")
 	BackupShard.Flags().Int32Var(&backupShardOptions.Concurrency, "concurrency", 4, "Specifies the number of compression/checksum jobs to run simultaneously.")
-	BackupShard.Flags().StringVar(&backupShardOptions.IncrementalFromPos, "incremental-from-pos", "", "Position of previous backup. Default: empty. If given, then this backup becomes an incremental backup from given position. If value is 'auto', backup taken from last successful backup position")
+	BackupShard.Flags().StringVar(&backupShardOptions.IncrementalFromPos, "incremental-from-pos", "", "Position, or name of backup from which to create an incremental backup. Default: empty. If given, then this backup becomes an incremental backup from given position or given backup. If value is 'auto', backup taken from last successful backup position.")
```
Comment (deepthi) [inline go/cmd/vtctldclient/command/backups.go:291]: Make this consistent with the other places.

File: go/vt/mysqlctl/builtinbackupengine.go [hunk 7]
```diff
@@ -291,6 +294,19 @@ func (be *BuiltinBackupEngine) executeIncrementalBackup(ctx context.Context, par
 	if err != nil {
 		return false, vterrors.Wrapf(err, "cannot get binary logs in incremental backup")
 	}
+
+	// @@gtid_purged
```
Comment (deepthi) [inline go/vt/mysqlctl/builtinbackupengine.go:298]: Is this comment line needed? Doesn't seem to add anything.

File: go/vt/vtctl/backup.go [hunk 8]
```diff
@@ -72,7 +72,7 @@ func init() {
 func commandBackup(ctx context.Context, wr *wrangler.Wrangler, subFlags *pflag.FlagSet, args []string) error {
 	concurrency := subFlags.Int32("concurrency", 4, "Specifies the number of compression/checksum jobs to run simultaneously")
 	allowPrimary := subFlags.Bool("allow_primary", false, "Allows backups to be taken on primary. Warning!! If you are using the builtin backup engine, this will shutdown your primary mysql for as long as it takes to create a backup.")
-	incrementalFromPos := subFlags.String("incremental_from_pos", "", "Position of previous backup. Default: empty. If given, then this backup becomes an incremental backup from given position. If value is 'auto', backup taken from last successful backup position")
+	incrementalFromPos := subFlags.String("incremental_from_pos", "", "Position, or name of backup from which to create an incremental backup. Default: empty. If given, then this backup becomes an incremental backup from given position or given backup. If value is 'auto', backup taken from last successful backup position.")
```
Comment (deepthi) [inline go/vt/vtctl/backup.go:75]: Make this consistent with all other flag docs.

File: go/vt/vtctl/backup.go [hunk 9]
```diff
@@ -114,7 +114,7 @@ func (b *backupEventStreamLogger) Send(resp *vtctldatapb.BackupResponse) error {
 func commandBackupShard(ctx context.Context, wr *wrangler.Wrangler, subFlags *pflag.FlagSet, args []string) error {
 	concurrency := subFlags.Int32("concurrency", 4, "Specifies the number of compression/checksum jobs to run simultaneously")
 	allowPrimary := subFlags.Bool("allow_primary", false, "Whether to use primary tablet for backup. Warning!! If you are using the builtin backup engine, this will shutdown your primary mysql for as long as it takes to create a backup.")
-	incrementalFromPos := subFlags.String("incremental_from_pos", "", "Position of previous backup. Default: empty. If given, then this backup becomes an incremental backup from given position. If value is 'auto', backup taken from last successful backup position")
+	incrementalFromPos := subFlags.String("incremental_from_pos", "", "Position, or name of backup from which to create an incremental backup. Default: empty. If given, then this backup becomes an incremental backup from given position or given backup. If value is 'auto', backup taken from last successful backup position.")
```
Comment (deepthi) [inline go/vt/vtctl/backup.go:117]: And this too.

**Review by deepthi — APPROVED** [review]
All my comments are minor. Please feel free to merge once they are addressed.

File: go/vt/vtctl/backup.go (hunk 8 above)
Comment (shlomi-noach) [reply go/vt/vtctl/backup.go]: Fixed.

File: go/vt/mysqlctl/builtinbackupengine.go (hunk 7 above)
Comment (shlomi-noach) [reply go/vt/mysqlctl/builtinbackupengine.go]: Removed.

File: go/vt/vtctl/backup.go (hunk 9 above)
Comment (shlomi-noach) [reply go/vt/vtctl/backup.go]: Fixed.

File: go/cmd/vtctldclient/command/backups.go (hunk 6 above)
Comment (shlomi-noach) [reply go/cmd/vtctldclient/command/backups.go]: Fixed.

**Pull request merged**